src/crm/activecampaign.ts

```diff
@@ -0,0 +1,300 @@
+};
+
+type ActiveCampaignPayload = {
+  contact: ActiveCampaignContact;
+};
+
+class ActiveCampaign extends CRM {
+  constructor(opt: {}) {
+    super(opt);
+    this.crmType = CRMType.ActionContact;
+  }
+
+   fetchCampaign = async (id: number): Promise<any> => {
+     const r = await procaCampaign(id);
+    return r;
+  }
+
+  headers: HeadersInit = {
+    'Api-Token': token!,
+    'Accept': 'application/json',
+    'Content-Type': 'application/json',
+  };
+
+  // actionid for the last campaign
+  // send data source, text field, value "proca"
+
+  body = ({ email, firstName, lastName, contactRef, id, phone, postcode }: ContactPayload,
+    data_source: string,
+    action_id_field: string,
+    ref_field: string,
+    zip_field: string) => {
+    const fieldValues = [
+      {
+        field: data_source, // data_source field ID
+        value: 'proca'
+      },
+      {
+        field: action_id_field,
+        value: id
+      },
+      {
+        field: ref_field,
+        value: contactRef
+      }
+    ];
+
+  // Add ZIP (postcode) only if it's provided
+  if (postcode) {
+    fieldValues.push({
+      field: zip_field, // ZIP field ID, default is 11
+      value: postcode
+    });
+  }
+
+  const contact: ActiveCampaignContact = {
+    firstName,
+    fieldValues
+  };
+
+  // Conditionally add optional fields
+  if (email) contact.email = email;
+  if (lastName) contact.lastName = lastName;
+  if (phone) contact.phone = phone;
+
+  return JSON.stringify({ contact });
+  };
+
+  async  getActiveCampaignFields() {
+    try {
+      const response = await fetch(`${url}/api/3/fields`, {
+        method: 'GET',
+        headers: this.headers
+      });
+      if (!response.ok) {
+        throw new Error(`Error fetching fields: ${response.status} ${response.statusText}`);
+      }
+      const data = await response.json();
+      console.log('Retrieved fields:', data.fields);
+      return data.fields;
+    } catch (error) {
+      console.error('Failed to fetch ActiveCampaign fields:', error);
+      return null;
+    }
+  }
+
+  async getActiveCampaignLists() {
+  try {
+    const response = await fetch(`${url}/api/3/lists`, {
+      method: 'GET',
+      headers: this.headers
+    });
+
+    if (!response.ok) {
+      throw new Error(`Error fetching lists: ${response.status} ${response.statusText}`);
+    }
+    const data = await response.json();
+    console.log('Retrieved lists:', data.lists);
+    return data.lists;
+  } catch (error) {
+    console.error('Failed to fetch ActiveCampaign lists:', error);
+    return null;
+  }
+}
+
+
+  fetchContact = async (email: string): Promise<string | null> => {
+    console.log("Fetching contact:", email);
+    try {
+      const res = await fetch(`${url}/api/3/contacts?email=${encodeURIComponent(email)}`, {
+        method: "GET",
+        headers: this.headers,
+      });
+
+      if (!res.ok) return null;
+
+      const data = await res.json();
+      return data.contacts?.[0]?.id || null;
```

Comment:
data.contacts?.[0]?.id returns undefined if the id isn't set, no need to || null, undefined is more semantically meaningful than null here I think

(and the way you use it later works the same if undefined or null)